File: ui/php/healthcheck/backend/checks/obm-sync/SyncStatus.php
<?php

/* ***** BEGIN LICENSE BLOCK *****
 * Copyright (C) 2011-2012  Linagora
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or (at your option) any
 * later version, provided you comply with the Additional Terms applicable for OBM
 * software by Linagora pursuant to Section 7 of the GNU Affero General Public
 * License, subsections (b), (c), and (e), pursuant to which you must notably (i)
 * retain the displaying by the interactive user interfaces of the “OBM, Free
 * Communication by Linagora” Logo with the “You are using the Open Source and
 * free version of OBM developed and supported by Linagora. Contribute to OBM R&D
 * by subscribing to an Enterprise offer !” infobox, (ii) retain all hypertext
 * links between OBM and obm.org, between Linagora and linagora.com, as well as
 * between the expression “Enterprise offer” and pro.obm.org, and (iii) refrain
 * from infringing Linagora intellectual property rights over its trademarks and
 * commercial brands. Other Additional Terms apply, see
 * <http://www.linagora.com/licenses/> for more details.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 * PARTICULAR PURPOSE. See the GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License and
 * its applicable Additional Terms for OBM along with this program. If not, see
 * <http://www.gnu.org/licenses/> for the GNU Affero General   Public License
 * version 3 and <http://www.linagora.com/licenses/> for the Additional Terms
 * applicable to the OBM software.
 * ***** END LICENSE BLOCK ***** */
 
require_once dirname(__FILE__) . '/../AbstractUserStatus.php';

class SyncStatus extends AbstractUserStatus {
  
  const SYNC_URL = "http://%HOST%:8080/obm-sync/services";
  
  public function executeForDomain($domain) {
    global $obm;

    $user = $this->getTestUser();
    $servers = of_domain_get_domain_syncserver($domain["id"]);
    
    foreach ($servers as $server) {
      $host = $server[0];      
      $url = str_replace("%HOST%", $host["ip"], self::SYNC_URL);
      $loginResult = self::callSync($url, "login/doLogin", array(
          "login" => $user['login'] . '@' . $domain['name'],
          "password" => $user['password'],
          "origin" => $user['origin']
      ));

      if (is_array($loginResult)) {
        return self::failure($domain['name'], $host['ip'], $loginResult);
      } else {
        $sid = (string) $loginResult->sid;
        $abSyncResult = self::callSync($url, "book/getAddressBookSync", array(
            "sid" => $sid,
            "lastSync" => time() * 1000 // Using current time to receive no changes/deletions
        ));
        
        if (is_array($abSyncResult)) {
          return self::failure($domain['name'], $host['ip'], $abSyncResult);
        }

        $obm['domain_id'] = $domain['id']; //So that get_user_id works as expected
        $userInfo = get_user_info(get_user_id($user['login']));

        $calSyncResult = self::callSync($url, "calendar/getSyncWithSortedChanges", array(
            "sid" => $sid,
            "calendar" => $userInfo['email'],
            "lastSync" => time() * 1000
        ));
        
        if (is_array($calSyncResult)) {
          return self::failure($domain['name'], $host['ip'], $calSyncResult);
        }
      }
      
      return new CheckResult(CheckStatus::OK);
    }
  }
  
  private static function failure($domain, $host, $messages) {
    return new CheckResult(CheckStatus::ERROR, array_merge(array("Synchronization failed for OBM-Sync server at '" . $host . "' for domain '" . $domain . "'."), $messages));
  }
  
  /**
   * Calls an obm-sync endpoint.
   * This returns either:
   * <ul>
   * <li>An array of strings if the call fails. The array holds all applicable error messages.</li>
   * <li>A SimpleXMLElement holding the returned XML data if the call succeeds</li>
   * </ul>
   * 
   * @param string $url Base obm-sync server URL
   * @param string $action obm-sync action, in the form of "handler/method".
   * @param array $params The parameters to send. This must be an assoc. array.
   * 
   * @return SimpleXMLElement|multitype:string
   */
  private static function callSync($url, $action, $params) {
    $messages = array();
    $curl = CheckHelper::curlGet($url . '/' . $action . '?' . http_build_query($params));
    
    if (!$curl['success'] || $curl['code'] != 200) {
      $messages[] = "Failed to invoke '" . $action . "' (Code: " . $curl['code'] . ")";
    } else {
      $xml = new SimpleXMLElement($curl['success']);
      
      if ($xml->getName() == "error") {
        $messages[] = "Failed to invoke '" . $action . "' (Error: " . $xml->message[0] . ")";
      } else {
        return $xml; // HTTP 200 and no <error />, call succeeded
      }
    }
    
    return $messages;
  }
  
}

File: java/sync/services-module/src/main/java/org/obm/sync/server/SyncStatus.java
/* ***** BEGIN LICENSE BLOCK *****
 * 
 * Copyright (C) 2011-2014  Linagora
 *
 * This program is free software: you can redistribute it and/or 
 * modify it under the terms of the GNU Affero General Public License as 
 * published by the Free Software Foundation, either version 3 of the 
 * License, or (at your option) any later version, provided you comply 
 * with the Additional Terms applicable for OBM connector by Linagora 
 * pursuant to Section 7 of the GNU Affero General Public License, 
 * subsections (b), (c), and (e), pursuant to which you must notably (i) retain 
 * the “Message sent thanks to OBM, Free Communication by Linagora” 
 * signature notice appended to any and all outbound messages 
 * (notably e-mail and meeting requests), (ii) retain all hypertext links between 
 * OBM and obm.org, as well as between Linagora and linagora.com, and (iii) refrain 
 * from infringing Linagora intellectual property rights over its trademarks 
 * and commercial brands. Other Additional Terms apply, 
 * see <http://www.linagora.com/licenses/> for more details. 
 *
 * This program is distributed in the hope that it will be useful, 
 * but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY 
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License 
 * for more details. 
 *
 * You should have received a copy of the GNU Affero General Public License 
 * and its applicable Additional Terms for OBM along with this program. If not, 
 * see <http://www.gnu.org/licenses/> for the GNU Affero General Public License version 3 
 * and <http://www.linagora.com/licenses/> for the Additional Terms applicable to 
 * OBM connectors. 
 * 
 * ***** END LICENSE BLOCK ***** */
package org.obm.sync.server;

import java.io.PrintWriter;
import java.util.Map;

import javax.servlet.http.HttpServletResponse;

import org.obm.sync.server.handler.ISyncHandler;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.google.inject.Inject;
import com.google.inject.Singleton;

import fr.aliacom.obm.common.ObmSyncVersion;
import fr.aliacom.obm.common.ObmSyncVersionNotFoundException;

@Singleton
public class SyncStatus {

	private static final String TITLE = "OBM Sync Status";

	private final Logger logger =  LoggerFactory.getLogger(getClass());

	private final Map<String, ISyncHandler> handlers;

	@Inject
	public SyncStatus(SyncHandlers handlers) {
		this.handlers = handlers.getHandlers();
	}

	public void show(HttpServletResponse resp) {
		resp.setContentType("text/html");
		try {

			PrintWriter w = resp.getWriter();

			w.println("<html><head><title>" + TITLE + "</title></head><body>");

			printVMInfo(w);

			printHandlers(w);

			printSyncInfo(w);

			w.print("</body></html>");

		} catch (Exception e) {
			logger.error("error showing status", e);
		}

	}

	private void printHandlers(PrintWriter w) {
		section("OBM Sync endpoints", w);
		for (String s : handlers.keySet()) {
			w.println("<p>handler configured for <em>" + s + "</em></p>");
		}
	}

	private void section(String string, PrintWriter w) {
		w.println("<h1>" + string + "</h1>");
	}

	private void printVMInfo(PrintWriter w) {
		section("Java Virtual Machine informations", w);
		w.println("<p><em>Vendor :</em> " + System.getProperty("java.vendor")
				+ "</p> ");
		w.println("<p><em>Version :</em> " + System.getProperty("java.version")
				+ "</p> ");
		w.println("<p><em>OS :</em> " + System.getProperty("os.name") + " "
				+ System.getProperty("os.version") + "</p> ");
		w.println("<p><em>Timezone :</em> "
				+ System.getProperty("user.timezone") + "</p> ");
		w.println("<p><em>Username :</em> " + System.getProperty("user.name")
				+ "</p> ");
	}

	private void printSyncInfo(PrintWriter w) {
		section("OBM Sync informations", w);
		w.println("<p><em>Version :</em> " + getObmSyncVersion() + "</p> ");
	}
	
	private String getObmSyncVersion() {
		try {
			return ObmSyncVersion.current().toString();
		} catch (ObmSyncVersionNotFoundException e) {
			return "Invalid obm-sync server version";
		}
	}

}
